bandwidth_group_mutators.go

```diff
@@ -0,0 +1,36 @@
+package transmissionrpc
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"reflect"
+)
+
+/*
+   Bandwidth Group Mutators
+   https://github.com/transmission/transmission/blob/4.0.2/docs/rpc-spec.md#481-bandwidth-group-mutator-group-set
+*/
+
+// BandwidthGroupSet applies a list of mutator(s) to a bandwidth group.
+func (c *Client) BandwidthGroupSet(ctx context.Context, payload BandwidthGroupSetPayload) (err error) {
+	// Validate
+	if reflect.ValueOf(payload.Name).Kind() == reflect.String {
```

Comment:
I think my intention here was to ensure the value of payload.Name is in fact a string. Checking for nil and empty string seems appropriate, since the actual idea is to ensure that at least a name is provided.

I don't remember the reason why it needs to be a pointer to a string -- I was basing it off of what hekmon was already doing in torrent_mutators.go. It's possible it would work better if the field were a string itself instead of a pointer to a string, but then maybe we might as well change all the fields to non-pointers? I'm not sure what hekmon's reasoning was for making them pointers. I'm also not terribly well-versed in Go, so I'd appreciate if you could inform me of a better way to do it.

Reply:
From a previous (and brief) conversation with hekmon in another thread he mentioned this library was one of his first and he's looking to improve it. My suggestion is to stop using pointers where the zero value doesn't make any difference, and that's probably "in most places." 

The reason we use a pointer in a struct when unmarshaling json is so the consumer of the unmarshalled json struct can determine if the field even existed in the json at all. If the field does not exist, the pointer will be nil after unmarshal. If the field does exist, the pointer will not be nil. Simple as that. When we're dealing with versioned APIs that have documented specifications, we can generally assume _all_ the fields will exist, and it's usually safe to assume the default zero value when the field does not exist. 

Nil pointers also behave differently than a zero-value field when marshaling a struct into json. You can make a zero-value behave like a nil pointer by suffixing `,omitempty` to the json tag name. I would bet that none of the fields needs to be pointers for any input (marshalled) struct.

It appears that `BandwidthGroupSetPayload` is an input struct. The optional fields can be pointers, but they don't need to be. The required fields should not be pointers. Then you just check for `payload.Name == ""` and error out as appropriate. But I wonder if you even need to validate that? Will Transmission validate it for you? Libraries should strive to be as dumb as possible, and rely on the API to produce appropriate errors for bad input. Then library maintainers get to send feedback to API maintainers so they can improve the API.

Reply:
For the use of pointers, I did not started the lib with it. But I soon encountered issue for setters where sending a default value (such as `0`) could be a valid command/param, therefor omitempty was not an option.

I could have used pointers only for them but I prefered to have one logic instead of a hybrid one with some pointers for some field and not for others.

I recall too that on some methods, the columns you ask in the query payload defines the columns you received in the anwer payload. Not having pointers on response payload could lead to usage a field value... that was not actually received.

About this specific test, I can not recall exactly, I need to re immerse me in the code.